framework/src/media/MediaPlayerImpl.cpp

```diff
@@ -146,96 +222,194 @@ player_result_t MediaPlayerImpl::unprepare()
-	mpw.getQueue().enQueue(&PlayerWorker::startPlayer, &mpw, shared_from_this());
+	mpw.getQueue().enQueue(&MediaPlayerImpl::startPlayer, shared_from_this());
 
 	return PLAYER_OK;
 }
 
+void MediaPlayerImpl::startPlayer()
+{
+	PlayerWorker& mpw = PlayerWorker::getWorker();
+
+	medvdbg("MediaPlayer Worker : startPlayer\n");
+	if (mCurState != PLAYER_STATE_READY && mCurState != PLAYER_STATE_PAUSED) {
+		notifyObserver(PLAYER_OBSERVER_COMMAND_ERROR);
+		return;
+	}
+
+	if (mpw.getPlayer() != nullptr) {
+		mpw.getPlayer()->pausePlayer();
+	}
+	mpw.setPlayer(shared_from_this());
+
+	mCurState = PLAYER_STATE_PLAYING;
+	notifyObserver(PLAYER_OBSERVER_COMMAND_STARTED);
+}
+
 player_result_t MediaPlayerImpl::stop()
 {
 	std::lock_guard<std::mutex> lock(mCmdMtx);
 	medvdbg("MediaPlayer stop\n");
-	if (mCurState == PLAYER_STATE_NONE || mCurState == PLAYER_STATE_IDLE) {
-		meddbg("MediaPlayer stop fail : wrong state\n");
+
+	PlayerWorker& mpw = PlayerWorker::getWorker();
+	if (!mpw.isAlive()) {
 		return PLAYER_ERROR;
 	}
 
-	PlayerWorker& mpw = PlayerWorker::getWorker();
-	mpw.getQueue().enQueue(&PlayerWorker::stopPlayer, &mpw, shared_from_this());
+	mpw.getQueue().enQueue(&MediaPlayerImpl::stopPlayer, shared_from_this());
 
 	return PLAYER_OK;
 }
 
+void MediaPlayerImpl::stopPlayer()
+{
+	PlayerWorker& mpw = PlayerWorker::getWorker();
+
+	medvdbg("MediaPlayer Worker : stopPlayer\n");
+	if (mCurState != PLAYER_STATE_PLAYING && mCurState != PLAYER_STATE_PAUSED) {
+		notifyObserver(PLAYER_OBSERVER_COMMAND_ERROR);
+		return;
+	}
+
+	mpw.setPlayer(nullptr);
+	mCurState = PLAYER_STATE_READY;
+}
+
 player_result_t MediaPlayerImpl::pause()
 {
 	std::lock_guard<std::mutex> lock(mCmdMtx);
 	medvdbg("MediaPlayer pause\n");
-	if (mCurState == PLAYER_STATE_NONE || mCurState == PLAYER_STATE_IDLE) {
-		meddbg("MediaPlayer pause fail : wrong state\n");
+
+	PlayerWorker& mpw = PlayerWorker::getWorker();
+	if (!mpw.isAlive()) {
 		return PLAYER_ERROR;
 	}
 
-	PlayerWorker& mpw = PlayerWorker::getWorker();
-	mpw.getQueue().enQueue(&PlayerWorker::pausePlayer, &mpw, shared_from_this());
+	mpw.getQueue().enQueue(&MediaPlayerImpl::pausePlayer, shared_from_this());
 
 	return PLAYER_OK;
 }
 
+void MediaPlayerImpl::pausePlayer()
+{
+	PlayerWorker& mpw = PlayerWorker::getWorker();
+
+	medvdbg("MediaPlayer Worker : pausePlayer\n");
+	if (mCurState != PLAYER_STATE_PLAYING) {
+		notifyObserver(PLAYER_OBSERVER_COMMAND_ERROR);
+		return;
+	}
+
+	mpw.setPlayer(nullptr);
+	mCurState = PLAYER_STATE_PAUSED;
+}
+
 int MediaPlayerImpl::getVolume()
 {
-	std::lock_guard<std::mutex> lock(mCmdMtx);
+	int ret = -1;
+
+	std::unique_lock<std::mutex> lock(mCmdMtx);
 	medvdbg("MediaPlayer getVolume\n");
+
+	PlayerWorker& mpw = PlayerWorker::getWorker();
+	if (!mpw.isAlive()) {
+		return PLAYER_ERROR;
+	}
+
+	mpw.getQueue().enQueue(&MediaPlayerImpl::getVolumePlayer, shared_from_this(), std::ref(ret));
+	mSyncCv.wait(lock);
+
+	return ret;
+}
+
+void MediaPlayerImpl::getVolumePlayer(int &ret)
```

Comment:
Why it changed void?

Reply:
There is `int MediaPlayerImpl::getVolume()` above

Reply:
Any thread get same result from audio manager through the getVolume, but it should be added to queue?
I mean atomicity is not needed for volume, because it is global value.....